[Citadel] Add all features to tpeplugin

claireyywang opened the pull request:
This PR add all the features to `tpeplugin` and enables testing with example worlds. This PR should be merged after #45 (adds needed functions to `tpelib`) and #32 (implements `Base` and `EntityManagementFeatures`). 

A few things need to be addressed by follow-up PRs

- Add collision detection
- Add a `VelocityCmd` to `ign-gazebo/src/systems/physics/Physics.cc` to enable setting world velocity without applying force to joint 

So far, any models without joint should be able to transform pose with TPE.

To test with ignition gazebo 

```bash
cd src/ign-gazebo
git checkout chapulina/tpe
cd ~/citadel_ws
colcon build --cmake-args -DBUILD_TESTING=OFF --merge-install --packages-up-to ignition-gazebo3
. ~/citadel_ws/install/setup.bash
ign gazebo -v 2 shapes.sdf --physics-engine ignition-physics-tpe-plugin
```
Terminal output
```bash
[GUI] [Wrn] [Application.cc:649] [QT] file::/ComponentInspector/ComponentInspector.qml:161: ReferenceError: ComponentsModel is not defined
construct model ground_plane
construct model box
construct model cylinder
construct model sphere
construct link link
construct link box_link
construct link cylinder_link
construct link sphere_link
construct collision collision
construct collision box_collision
construct collision cylinder_collision
construct collision sphere_collision
```

The shapes no longer collapse to origin after the commit https://github.com/ignitionrobotics/ign-physics/pull/32/commits/344f7f48cd16841a5d683c43a72706fce28a54c4 , and warnings are fixed by commit https://github.com/ignitionrobotics/ign-physics/pull/32/commits/01bf3deacd965724a19d1ede8ca1333f5541f18a 
Shapes can be moved around by using transform tools after commit https://github.com/ignitionrobotics/ign-physics/pull/32/commits/ad391b3b4eb6a0299b3bf59ba9ca76aded48bc69

Demo
![TPE_shapes_demo](https://user-images.githubusercontent.com/22240514/80547524-4af17480-896d-11ea-9181-5d5705e7d0e0.gif)

I also tried with a couple other example worlds. See performance summary below:

- `actor.sdf`: loads fine, actors move, warns missing joint entity, can't transform pose
- `apply_joint_force.sdf`: loads fine, warns missing joint entity, can't transform pose
- `breadcrumbs.sdf`: loads fine, warns missing joint, can't transform pose
- `camera_sensor.sdf`: loads fine, sensor works, cube transforms pose, cone doesn’t and warns missing joint entity
- `contact_sensor`: loads fine, transforms pose, sensor works
- `fuel.sdf`: loads fine, warns missing joint entity, actor can move, can't transform pose
- `tunnel.sdf`: loads fine, warns missing joint entity, car transforms pose

chapulina commented on tpe/plugin/src/SimulationFeatures_TEST.cc:
```diff
@@ -0,0 +1,149 @@
+  if (library.empty())
+    return;
+
+  auto worlds = LoadWorlds(library, TEST_WORLD_DIR "/falling.world");
```
I think this world needs to be copied over from `dartsim`.

Or even better, moved to a common location so both plugins can use the same files in their tests.

chapulina commented on tpe/plugin/src/SimulationFeatures_TEST.cc:
```diff
@@ -0,0 +1,149 @@
+}
+
+INSTANTIATE_TEST_CASE_P(PhysicsPlugins, SimulationFeatures_TEST,
+  ::testing::ValuesIn(ignition::physics::test::g_PhysicsPluginLibraries),); // NOLINT
```
It looks like this test is not being run for TPE, because TPE is not listed on `g_PhysicsPluginLibraries`. :sweat_smile: 

You can add it like this:

```.diff
diff --git a/test/PhysicsPluginsList.hh b/test/PhysicsPluginsList.hh
index 7d047eb..99c80f6 100644
--- a/test/PhysicsPluginsList.hh
+++ b/test/PhysicsPluginsList.hh
@@ -29,6 +29,10 @@
 #define bullet_plugin_LIB ""
 #endif
 
+#ifndef tpe_plugin_LIB
+#define tpe_plugin_LIB ""
+#endif
+
 namespace ignition
 {
   namespace physics
@@ -38,6 +42,7 @@ namespace ignition
       const std::vector<std::string> g_PhysicsPluginLibraries = {
         dartsim_plugin_LIB,
         bullet_plugin_LIB,
+        tpe_plugin_LIB
       };
     }
   }
```

But then this will make a lot of other tests run for both DART and TPE. So shouldn't we just put these test in a common place? Right now this file is mostly a copy of `dartsim/src/SimulationFeatures_TEST.cc`.

claireyywang commented:
A couple note about recent commits to this PR: 

- addressed most of @chapulina 's review comments. As per offline discussion, I think it's better to make consolidating test files and world files for DART and TPE a separate task, and this PR will just focus on getting TPE tested and merged. 
- resolved some windows/macOS CI warnings
- added more test cases for all features, but unsure how to test `mesh` related functions.
- codecov diff mostly comes from `ignwarn`, `igndbg` and `ignerr` print statements and `mesh` functions

claireyywang commented on tpe/plugin/src/SDFFeatures_TEST.cc:
```diff
@@ -0,0 +1,333 @@
+using World = ignition::physics::World3d<TestFeatureList>;
+using WorldPtr = ignition::physics::World3dPtr<TestFeatureList>;
+
+auto LoadEngine()
```
I don't think any of dartsim's test files has it, and am leaning towards keeping it consistent.

iche033 commented on tpe/plugin/src/SDFFeatures.cc:
```diff
@@ -0,0 +1,219 @@
+{
+  // Read sdf params
+  const std::string name = _sdfModel.Name();
+  igndbg << "construct model " << name << std::endl;
```
I think I may have added these. We can remove them now that there tests verifying the entities are created.

iche033 commented on tpe/plugin/src/ShapeFeatures.cc:
```diff
@@ -0,0 +1,309 @@
+  }
+  // return invalid bounding box if collision not found
+  return math::eigen3::convert(
+    math::AxisAlignedBox(
```
how about just `math::AxisAlignedBox()` with its default [min / max](https://github.com/ignitionrobotics/ign-math/blob/master/src/AxisAlignedBox.cc#L27) values. We have been using that to represent an invalid bounding box

claireyywang commented:
I cleaned up the code a bit more. Ready for another round of review :bowing_woman:

iche033 commented:
I made some very minor changes in f952393084d03dd97eaf6212d3f3d31cb0d4effa. I think the PR is good to go. Approve once the homebrew issue is fixed.

claireyywang commented:
> I made some very minor changes in [f952393](https://github.com/ignitionrobotics/ign-physics/commit/f952393084d03dd97eaf6212d3f3d31cb0d4effa). I think the PR is good to go. Approve once the homebrew issue is fixed.

I think the homebrew issue was due to a Java bug on `ultron.mojave`. The CI was rebuilt on another machine and shows green now.

claireyywang commented:
@chapulina mind taking another look since there are requested changes from you blocking merge? Thanks!

chapulina commented on tpe/plugin/src/SimulationFeatures_TEST.cc:
```diff
@@ -0,0 +1,240 @@
+
+    auto sphere2 = sphereLink->AttachSphereShape(
+      "sphere2", 1.0, Eigen::Isometry3d::Identity());
+    EXPECT_EQ(2u, sphereLink->GetShapeCount());
```
Should we get the count before adding too? To check that it increased?

chapulina commented on tpe/plugin/src/SimulationFeatures_TEST.cc:
```diff
@@ -0,0 +1,240 @@
+  {
+    auto model = world->GetModel("sphere");
+    auto freeGroup = model->FindFreeGroup();
+    EXPECT_NE(nullptr, freeGroup);
```
This should be `ASSERT_NE`, so if the check fails, the test ends here instead of crashing on the line below.

claireyywang commented on tpe/plugin/src/KinematicsFeatures.cc:
```diff
@@ -0,0 +1,67 @@
+  auto it = this->childIdToParentId.find(_id.ID());
+  if (it == this->childIdToParentId.end())
+  {
+    ignwarn << "Link [" << _id.ID() << "]  is not found." << std::endl;
```
That's a good point. I think back then I was only considering the simplest use case. I added world entity to the function. Now users can get framedata of link relative to model or framedata of model relative to world.

chapulina replied:
> Now users can get framedata of link relative to model or framedata of model relative to world.

I'm not sure I understand, you mean with the current implementation of `FrameDataRelativeToWorld`? I think the result should always be with respect to the world, right?

claireyywang replied:
oops sorry, I meant to say the frameID now can refer to a link or a model, and yes result is always relative to world :sweat_smile:

chapulina replied:
I think you addressed this on https://github.com/ignitionrobotics/ign-physics/pull/46/commits/4729d5fc8b818f05dac929035b5c0b6bb9bad2c7 but I don't see that code in the current branch. Bad merge?

claireyywang replied:
interesting... it showed up on mine

claireyywang replied:
https://github.com/ignitionrobotics/ign-physics/pull/46/files#diff-113395042bd9e76d68081a83d3bfa56aR49 here?

chapulina replied:
ha, ok sorry, I just don't know how to use GitHub :sweat_smile: I thought that clicking on the file name up here would take me to the latest version of that file. Oh gosh, I must have been giving a lot of weird feedback lately :laughing:

claireyywang commented on tpe/plugin/src/KinematicsFeatures.cc:
```diff
@@ -0,0 +1,67 @@
+
+  data.pose = math::eigen3::convert(model->GetPose());
+  data.linearVelocity = math::eigen3::convert(model->GetLinearVelocity());
+  data.angularVelocity = math::eigen3::convert(model->GetAngularVelocity());
```
I think the assumption here is the pose of link and collision is always consistent with model and by setting the velocity and pose of a model, we set them for all entities inside the model. Tbh I'd rather implement the offset in a separate PR if that's deemed necessary/need arises since this PR is already quite large.

chapulina replied:
Ok, we can leave this for later since this PR is quite large already. My worry is that we can easily forget to come back to this and end up having to hunt this problem in the future. I think any multi-link model would be broken with the current implementation, because I believe such models will always have link offsets.

How about:

* Writing a TODO saying that the link pose needs to be considered
* Ticketing an issue in case we're not planning to tackle this next
* Bonus: detecting when the link pose is not zero and printing a warning (at this point, it may be just a bit more effort to add the link offset to these results)

claireyywang replied:
Cool, I have added a TODO https://github.com/ignitionrobotics/ign-physics/pull/46/commits/5215a743061e5a6a54dc193726a50f5cd1ff9480 and ticketed an issue https://github.com/ignitionrobotics/ign-physics/issues/57

claireyywang commented on tpe/plugin/src/SDFFeatures.cc:
```diff
@@ -0,0 +1,222 @@
+}
+
+/////////////////////////////////////////////////
+Identity SDFFeatures::ConstructSdfCollision(
```
I'd like to think there's function overloading happening, but you might be right. The problem here is that mesh has to be passed in as a parameter and doing so violates the `ConstructSdfCollision` [function template](https://github.com/ignitionrobotics/ign-physics/blob/master/sdf/include/ignition/physics/sdf/ConstructCollision.hh#L42) I'm not sure what's the best thing to do here. I guess I can remove the function for now if the support for mesh is not definite. @iche033 what do you think?

chapulina replied:
I think you could add an `else if (geom->Type() == ::sdf::GeometryType::MESH)` to the other `ConstructSdfCollision`. Then the question is how to go from `_sdfCollision->MeshShape()` to `common::Mesh`.

Or if meshes are not an important use case for TPE, this could be removed for now.

iche033 replied:
it may be fine to remove it. I may have added it think that it's necessary but `dartsim` also does not implement this. We can always add it back if needed

iche033 replied:
actually `dartsim` does have the `ConstructSdfCollision` function implemented but it's just not used by ign-gazebo. There is also a [TODO note](https://github.com/ignitionrobotics/ign-physics/blob/master/dartsim/src/SDFFeatures.cc#L291) for constructing meshes so I'm not sure if that function works for meshes or not.

We could either remove ConstructSdfCollision for tpe or just do what dartsim does and ignore meshes for now.

claireyywang replied:
I did what dartsim does and removed the function for meshes https://github.com/ignitionrobotics/ign-physics/pull/46/commits/8f58c69cfdae5c0e0d83f3f3c99e9a2cc349d939 also kept the TODO

iche033 replied:
lgtm
